ENIGMAsystem/SHELL/Universal_System/Serialization/serialization.h

```diff
@@ -0,0 +1,817 @@
+inline std::array<std::byte, sizeof(T)> internal_serialize_integral(T value) {
+  return serialize_any<std::make_unsigned_t<T>>(value);
+}
+
+template <typename T>
+inline void internal_serialize_floating_into(std::byte *iter, T value) {
+  if constexpr (std::is_same_v<T, float>) {
+    internal_serialize_any_into<std::uint32_t>(iter, value);
+  } else if constexpr (std::is_same_v<T, double>) {
+    internal_serialize_any_into<std::uint64_t>(iter, value);
+  } else {
+    static_assert(always_false<T>, "'internal_serialize_floating_into' only accepts 'float' or 'double' types");
+  }
+}
+
+template <typename T>
+inline std::array<std::byte, sizeof(T)> internal_serialize_floating(T value) {
+  if constexpr (std::is_same_v<T, float>) {
+    return serialize_any<std::uint32_t>(value);
+  } else if constexpr (std::is_same_v<T, double>) {
+    return serialize_any<std::uint64_t>(value);
+  } else {
+    static_assert(always_false<T>, "'internal_serialize_floating' only accepts 'float' or 'double' types");
+  }
+}
+
+template <typename T>
+inline void internal_serialize_numeric_into(std::byte *iter, T value) {
+  if constexpr (std::is_integral_v<T>) {
+    internal_serialize_integral_into(iter, value);
+  } else if constexpr (std::is_floating_point_v<T>) {
+    internal_serialize_floating_into(iter, value);
+  } else {
+    static_assert(always_false<T>, "'internal_serialize_numeric_into' takes either integral or floating types");
+  }
+}
+
+template <typename T>
+inline std::array<std::byte, sizeof(T)> internal_serialize_numeric(T value) {
+  if constexpr (std::is_integral_v<T>) {
+    return internal_serialize_integral(value);
+  } else if constexpr (std::is_floating_point_v<T>) {
+    return internal_serialize_floating(value);
+  } else {
+    static_assert(always_false<T>, "'internal_serialize_numeric' takes either integral or floating types");
+  }
+}
+
+template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+inline T internal_deserialize_integral(std::byte *iter) {
+  return internal_deserialize_any<std::make_unsigned_t<T>, T>(iter);
+}
+
+template <typename T>
+inline T internal_deserialize_floating(std::byte *iter) {
+  if constexpr (std::is_same_v<T, float>) {
+    return internal_deserialize_any<std::uint32_t, T>(iter);
+  } else if constexpr (std::is_same_v<T, double>) {
+    return internal_deserialize_any<std::size_t, T>(iter);
+  } else {
+    static_assert(always_false<T>, "'internal_deserialize_floating' only accepts 'float' or 'double' types");
+  }
+}
+
+template <typename T>
+inline T internal_deserialize_numeric(std::byte *iter) {
+  if constexpr (std::is_integral_v<T>) {
+    return internal_deserialize_integral<T>(iter);
+  } else if constexpr (std::is_floating_point_v<T>) {
+    return internal_deserialize_floating<T>(iter);
+  } else {
+    static_assert(always_false<T>, "'internal_deserialize_numeric' takes either integral or floating types");
+  }
+}
```

Comment:
Going forward, I think it may be a good idea to either:
- have six separate files for the six functions where each file contains implementations for all types
- have (roughly) separate files for each type implementing all of the functions per file

Of these, I like the second option more because it groups all the functionality for either a type or a set of related types together without noise from implementations for other types. In this case, I think it'd be a good idea to have all the functions for the integral and floating types in one file.